server/router/stats.ts

```diff
@@ -0,0 +1,155 @@
+    dates.push(date);
+    date = moment.tz(date, 'America/New_York').add(1, 'days').format('YYYY-MM-DD');
+  }
+
+  dates.forEach(async (currDate) => {
+    const year = moment.tz(currDate, 'YYYY-MM-DD', 'America/New_York').format('YYYY');
+    const monthDay = moment.tz(currDate, 'YYYY-MM-DD', 'America/New_York').format('MMDD');
+    const condition = new Condition()
+      .where('year')
+      .eq(year)
+      .and()
+      .where('monthDay')
+      .eq(monthDay);
+
+    const dateMoment = moment.tz(currDate, 'America/New_York');
+    // day = 12am to 5:00pm
+    const dayStart = dateMoment.toISOString();
+    const dayEnd = dateMoment.add(17, 'hours').toISOString();
+    // night = 5:01pm to 11:59:59pm
+    const nightStart = moment.tz(dayEnd, 'America/New_York').add(1, 'seconds').toISOString();
+    const nightEnd = moment.tz(currDate as string, 'America/New_York').endOf('day').toISOString();
+
+    computeStats(res, condition, dayStart, dayEnd, nightStart, nightEnd, year,
+      monthDay, currDate, to);
+  });
+});
+
+function computeStats(
+  res: Response,
+  condition: Condition,
+  dayStart: string,
+  dayEnd: string,
+  nightStart: string,
+  nightEnd: string,
+  year: string,
+  monthDay: string,
+  date: string,
+  to: string,
+) {
+  db.scan(res, Stats, condition, async (data: Document[]) => {
+    if (data.length) {
+      res.write(JSON.stringify(data));
+      if (date === to) {
+        res.end();
+      }
+    } else {
+      const conditionRidesDay = new Condition()
+        .where('startTime')
+        .between(dayStart, dayEnd)
+        .where('type')
+        .not()
+        .eq('unscheduled')
+        .where('status')
+        .eq('completed');
+
+      const conditionRidesDayNoShow = new Condition()
+        .where('startTime')
+        .between(dayStart, dayEnd)
+        .where('type')
+        .not()
+        .eq('unscheduled')
+        .where('status')
+        .eq('no_show');
+
+      const conditionRidesNight = new Condition()
+        .where('startTime')
+        .between(nightStart, nightEnd)
+        .where('type')
+        .not()
+        .eq('unscheduled')
+        .where('status')
+        .eq('completed');
+
+      const conditionRidesNightNoShow = new Condition()
+        .where('startTime')
+        .between(nightStart, nightEnd)
+        .where('type')
+        .not()
+        .eq('unscheduled')
+        .where('status')
+        .eq('no_show');
+
+      db.scan(res, Ride, conditionRidesDay, ((dataDay: Document[]) => {
+        const dayCountStat = dataDay.length;
+
+        db.scan(res, Ride, conditionRidesDayNoShow, ((dataDayNoShow: Document[]) => {
+          const dayNoShowStat = dataDayNoShow.length;
+
+          db.scan(res, Ride, conditionRidesNight, ((dataNight: Document[]) => {
+            const nightCountStat = dataNight.length;
+
+            db.scan(res, Ride, conditionRidesNightNoShow, ((dataNightNoShow: Document[]) => {
+              const nightNoShowStat = dataNightNoShow.length;
```

Comment:
to reduce database reads, could you just get all rides w/ stats not generated between `to` and `from`, then do all the filtering with that?